test/test_main.py

```diff
@@ -497,5 +458,105 @@ def test_logger(self):
+
+        for initial_event in initial_events:
+            self._setup_charm_dir()
+
+            self._simulate_event(initial_event)
+            _assess_event_links(initial_event)
+            # Make sure it is idempotent.
+            self._simulate_event(initial_event)
+            _assess_event_links(initial_event)
+
+    def test_setup_action_links(self):
+        charm_config = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+        }))
+        actions_yaml = self.JUJU_CHARM_DIR / 'actions.yaml'
+        actions_yaml.write_text('test: {}')
+        self._simulate_event(EventSpec(InstallEvent, 'install', charm_config=charm_config))
+        action_hook = self.JUJU_CHARM_DIR / 'actions' / 'test'
+        self.assertTrue(action_hook.exists())
+
+
+class TestMainWithDispatch(TestMain, unittest.TestCase):
+    def _symlink_entry_point(self, directory, entry_point):
+        path = self.JUJU_CHARM_DIR / 'dispatch'
+        if not path.exists():
+            path.symlink_to(self.JUJU_CHARM_DIR / 'src/charm.py')
+
+    def _call_event(self, rel_path, env):
+        env["JUJU_DISPATCH_PATH"] = str(rel_path)
+        dispatch = self.JUJU_CHARM_DIR / 'dispatch'
+        subprocess.check_call([sys.executable, str(dispatch)],
+                              env=env, cwd=str(self.JUJU_CHARM_DIR))
+
+    def test_no_setup_event_links(self):
+        """Test auto-creation of symlinks caused by initial events does _not_ happen when using dispatch.
+        """
+        all_event_hooks = ['hooks/' + e.replace("_", "-")
+                           for e in self.charm_module.Charm.on.events().keys()]
+        charm_config = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+        }))
+        initial_events = {
+            EventSpec(InstallEvent, 'install', charm_config=charm_config),
+            EventSpec(StorageAttachedEvent, 'disks-storage-attached', charm_config=charm_config),
+            EventSpec(StartEvent, 'start', charm_config=charm_config),
+            EventSpec(UpgradeCharmEvent, 'upgrade-charm', charm_config=charm_config),
+        }
+
+        def _assess_event_links(event_spec):
+            self.assertNotIn(self.hooks_dir / event_spec.event_name, self.hooks_dir.iterdir())
+            for event_hook in all_event_hooks:
+                self.assertFalse((self.JUJU_CHARM_DIR / event_hook).exists(),
+                                 'Spurious hook: ' + event_hook)
+
+        for initial_event in initial_events:
+            self._setup_charm_dir()
+
+            self._simulate_event(initial_event)
+            _assess_event_links(initial_event)
+
+    def test_hook_overrides_dispatch(self):
+        charm_config = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+        }))
+
+        self._simulate_event(EventSpec(InstallEvent, 'install', charm_config=charm_config))
+
+        path = self.hooks_dir / 'install'
+        path.symlink_to('/bin/false')
```

Comment:
I think we'll need something a bit more here, so that we can see that both we do invoke 'install' but also we trigger our internal handlers.